compiler/test/compilable/test23708.d

```diff
@@ -0,0 +1,10 @@
+// https://issues.dlang.org/show_bug.cgi?id=23708
+
+// REQUIRED_ARGS: -preview=nosharedaccess
+
+class Class {}
+shared(Class) fun()
+{
+    static shared Class ret;
+    return ret;
```

Comment:
I don't understand, why are classes special here? What is preventing another thread from modifying `ret` at the same time as a non-atomic return?

Reply:
This code is simply passing `ret` to the caller, it's not modifying it in any way.

Reply:
This code doesn't actually share `ret` either, but that's not the point. Access to `shared` variables in general is forbidden to avoid race conditions. The compiler rejects the function with any other type: `ulong`, `int*`, `string`... why are classes exempt from this here?

Reply:
Because this returning references works:

```d
ref shared(int) fun()                                                                                                                                                                                                                            
{
    static shared int ret;
    return ret;
}
```

Reply:
There is no documentation for -preview=nosharedaccess and the DIP is lacking in details in my opinion. However, in this context I am interpreting shared as actually sharing the data, not the references to the shared data. Ideally, passing references to shared data should not be subject to `shared` restrictions.

Reply:
"Because you can return a `shared(int)` by `ref`, reading a shared class variable cannot result in a race condition."
That makes no sense.

Reply:
You can also return a `shared(Class)` by `ref`, returning by `ref` doesn't read the variable, it returns a pointer to the variable.

Reply:
But you are not actually reading the class contents, you are just reading the reference to the memory region that holds the class. The same as when returning a reference to an `int`.

Reply:
If that's the case, then a `shared class` should be equivalent to a `shared(classInstance)*` instead of a `shared(classInstance*)`. That's not currently the case however:
```D
Object oTls; // put in tls
shared Object oShared; // put in shared data segment
```

Not only the class contents, but also the pointer value stored in `oShared` can be mutated from any thread, so you cannot return it directly, it might be in a corrupt state in the middle of another thread's write to it.

Reply:
How can the pointer value be mutated?

Reply:
```D
oShared = new Object();
```

Reply:
More complete example:
```D
shared Object oShared;
void main()
{
    import core.thread;
    auto t = new Thread(() {
        while (1) {
            // lock mutex
            *(cast(Object*) &oShared) = new Object();
            // unlock mutex
        }
    }).start();
}
```

Reply:
Yes, but in the context of the issue reported by @atilaneves we are talking about returning a copy of the reference, so I would assume that if you are fine with getting a snapshot of the object at any given moment in time that should not be a problem. Even if you would use a mutex, there is no guarantee that you are going to get a reference to the older object or the newer one.

I see what you mean, but from my perspective a case could be made for both options. @atilaneves what do you think?

Reply:
> Even if you would use a mutex, there is no guarantee that you are going to get a reference to the older object or the newer one.

Yes, but at least there's a guarantee that you're not getting a corrupted pointer that's 32 bits old and 32 bits new. 

> a case could be made for both options

No! If you're allowed to produce garbage pointers just like that it completely defeats the point of `@safe` and `shared`.

Reply:
> Yes, but at least there's a guarantee that you're not getting a corrupted pointer that's 32 bits old and 32 bits new.

Is that even possible? I would expect that that is impossible on modern architectures.

Reply:
Well, technically it is possible, it depends on the platform.

Reply:
Just because it might happen to work out on a modern architecture doesn't mean DMD can break its own type system.

And if we're going to assume that 8-byte reads are always atomic, then this fix should not be limited to classes, but also work on integers, floats, and other pointers. 

Note that it's not just about whether the instructions are atomic either, there's also memory order and sequential consistency to consider.